src/metatrain/experimental/nanopet/model.py

```diff
@@ -0,0 +1,474 @@
+            )
+
+        # register new outputs as new last layers
+        for target_name, target in new_targets.items():
+            self._add_output(target_name, target)
+
+        self.dataset_info = merged_info
+        self.atomic_types = sorted(self.atomic_types)
+
+        return self
+
+    def forward(
+        self,
+        systems: List[System],
+        outputs: Dict[str, ModelOutput],
+        selected_atoms: Optional[Labels] = None,
+    ) -> Dict[str, TensorMap]:
+        # Checks on systems (species) and outputs are done in the
+        # MetatensorAtomisticModel wrapper
+
+        device = systems[0].device
+
+        if self.single_label.values.device != device:
+            self.single_label = self.single_label.to(device)
+            self.key_labels = {
+                output_name: label.to(device)
+                for output_name, label in self.key_labels.items()
+            }
+            self.component_labels = {
+                output_name: [label.to(device) for label in labels]
+                for output_name, labels in self.component_labels.items()
+            }
+            self.property_labels = {
+                output_name: label.to(device)
+                for output_name, label in self.property_labels.items()
+            }
+
+        segment_indices = torch.concatenate(
+            [
+                torch.full(
+                    (len(system),),
+                    i_system,
+                    device=device,
+                )
+                for i_system, system in enumerate(systems)
+            ],
+        )
+
+        sample_values = torch.stack(
+            [
+                segment_indices,
+                torch.concatenate(
+                    [
+                        torch.arange(
+                            len(system),
+                            device=device,
+                        )
+                        for system in systems
+                    ],
+                ),
+            ],
+            dim=1,
+        )
+        sample_labels = Labels(
+            names=["system", "atom"],
+            values=sample_values,
+        )
+
+        (
+            positions,
+            centers,
+            neighbors,
+            species,
+            cells,
+            cell_shifts,
+        ) = concatenate_structures(systems)
+
+        # somehow the backward of this operation is very slow at evaluation,
+        # where there is only one cell, therefore we simplify the calculation
+        # for that case
+        if len(cells) == 1:
+            cell_contributions = cell_shifts.to(cells.dtype) @ cells[0]
+        else:
+            cell_contributions = torch.einsum(
+                "ab, abc -> ac",
+                cell_shifts.to(cells.dtype),
+                cells[segment_indices[centers]],
+            )
+
+        edge_vectors = positions[neighbors] - positions[centers] + cell_contributions
```

Comment:
why are you re-computing the edge vectors (and adding a bunch of nodes to the backward) instead of using the neighbor list values?

Reply:
Because it's cleaner and faster. I profiled it and we also talked about it (no difference at evaluation but faster at training)

Reply:
I'm a bit surprised that this is faster, given that this is adding more redundant computations. But we can keep it like this for now & revisit later as needed.

Reply:
We can discuss in person